File: 0133-clone-graph/0133-clone-graph.py
"""
# Definition for a Node.
class Node(object):
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []
"""

class Solution(object):
    def cloneGraph(self, node):
        """
        :type node: Node
        :rtype: Node
        """
        if not node:
            return node
        ans, visited, q = {}, [], deque([])
        q.append(node)
        while q:
            v = q.popleft()
            if v not in visited:
                visited.append(v)
                if v not in ans:
                    ans[v] = Node(v.val)
                for w in v.neighbors:
                    q.append(w)
                    if w not in ans:
                        ans[w] = Node(w.val)
                    ans[v].neighbors.append(ans[w])
        return ans[node]        

File: 0133-clone-graph/0133-clone-graph.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> neighbors;
    Node() {
        val = 0;
        neighbors = vector<Node*>();
    }
    Node(int _val) {
        val = _val;
        neighbors = vector<Node*>();
    }
    Node(int _val, vector<Node*> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
};
*/

class Solution {
public:
    Node* dfs(Node* n, unordered_map<Node*, Node*>& graph) {
        Node* copied_n = new Node(n->val);
        vector<Node*> copied_neighbors;
        graph[n] = copied_n;
        for (auto nei : n->neighbors) {
            if (graph.find(nei) != graph.end()) {
                copied_neighbors.push_back(graph[nei]);
            } else {
                // Node* nb = new Node(nei->val);
                copied_neighbors.push_back(dfs(nei, graph));
            }
        }

        copied_n->neighbors = copied_neighbors;
        return copied_n;
    }

    Node* cloneGraph(Node* node) {
        if (!node) return NULL;
        if (node->neighbors.size()==0) {
            Node* copied = new Node(node->val);
            return copied;
        }

        unordered_map<Node*, Node*> ans;
        return dfs(node, ans);
    }
};